### PR TITLE
tests, some fixes

### PR DESCRIPTION
**Pull request by Loloxon**

### CODE REVIEW

File: graph/elements/node.py [hunk 1]
```diff
@@ -15,5 +15,5 @@ def __init__(self, x: float, y: float, is_hanging: bool, label: str = "V"):
         self.label = label
 
     def __str__(self):
-        string = f"Node: {self.x} {self.y}, {self.is_hanging} ({self.label})"
+        string = f"Node: {self.x}.{self.y}, {self.is_hanging} ({self.label})"
```
Comment (strjakub) [inline graph/elements/node.py:18]: I removed it deliberately, debugging with this dot is fucked up, you see coords like this `0.0.1.0`, it took me 10 minutes to understand what happens there

File: graph/hypergraph.py [hunk 2]
```diff
@@ -196,16 +196,19 @@ def parse_networkx_to_hypergraph(nx_graph: nx.Graph) -> 'HyperGraph':
                     if not isinstance(neighbor, str):  # Regular node (not a hyperedge node)
                         neighbor_obj = next(
                             n for n in nodes if
-                                n.x == nx_graph.nodes[neighbor]['x'] and n.y == nx_graph.nodes[neighbor]['y']
-                            )
+                            n.x == nx_graph.nodes[neighbor]['x'] and n.y == nx_graph.nodes[neighbor]['y']
+                        )
```
Comment (strjakub) [inline graph/hypergraph.py:200]: why change this? it was more readable

File: productions/p1.py [hunk 3]
```diff
@@ -61,17 +61,17 @@ def _right_side(self) -> HyperGraph:
...
             reversed_node_map = {v: k for k, v in node_map.items() if isinstance(v, Node) and isinstance(k, Node)}
 
-            node_1 = graph.split_edge_if_exist(reversed_node_map[left_side.nodes[0]], reversed_node_map[left_side.nodes[1]], False)
-            node_2 = graph.split_edge_if_exist(reversed_node_map[left_side.nodes[1]], reversed_node_map[left_side.nodes[2]], False)
-            node_3 = graph.split_edge_if_exist(reversed_node_map[left_side.nodes[2]], reversed_node_map[left_side.nodes[3]], False)
-            node_4 = graph.split_edge_if_exist(reversed_node_map[left_side.nodes[3]], reversed_node_map[left_side.nodes[0]], False)
+            node_1 = graph.split_edge_if_exist(reversed_node_map[left_side.nodes[0]], reversed_node_map[left_side.nodes[1]], True)
+            node_2 = graph.split_edge_if_exist(reversed_node_map[left_side.nodes[1]], reversed_node_map[left_side.nodes[2]], True)
+            node_3 = graph.split_edge_if_exist(reversed_node_map[left_side.nodes[2]], reversed_node_map[left_side.nodes[3]], True)
+            node_4 = graph.split_edge_if_exist(reversed_node_map[left_side.nodes[3]], reversed_node_map[left_side.nodes[0]], True)
```
Comment (strjakub) [inline productions/p1.py:74]: it depends on `edge.is_border`

File: productions/p1.py [hunk 4]
```diff
@@ -86,4 +86,5 @@ def apply(self, graph: HyperGraph) -> HyperGraph:
             graph.add_hyperedge([node_3, node_4, reversed_node_map[left_side.nodes[3]], node_center])
             graph.add_hyperedge([node_4, node_1, reversed_node_map[left_side.nodes[0]], node_center])
 
-        return graph
+            return graph
+        return None
```
Comment (strjakub) [inline productions/p1.py:90]: I am not fan of that, in the future we will need to pipe graph through production after production. It is reasonable to return graph that can go straight into next production

File: productions/production.py [hunk 5]
```diff
@@ -25,7 +25,7 @@ def _right_side(self) -> HyperGraph:
         pass
 
     @abstractmethod
-    def apply(self, graph: HyperGraph) -> HyperGraph:
+    def apply(self, graph: HyperGraph) -> HyperGraph | None:
```
Comment (strjakub) [inline productions/production.py:28]: as described in implementation of P1

**Review by strjakub — CHANGES_REQUESTED** [review]
I will change production abstract methods tomorrow or Friday morning to extract `check` logic from apply. I think of making apply not abstract but using abstract `transform` inside `if` that is currently implemented in P1 will be the best. Then everyone would only need to implement production changes having dictionary that map nodes of `left_side` to real `graph` nodes

File: graph/hypergraph.py [hunk 6]
```diff
@@ -27,30 +27,14 @@ def print(self):
...
+    def get_edge_between(self, node_1: Node, node_2: Node):
         for edge in self.edges:
             if (edge.node_1 == node_1 and edge.node_2 == node_2) or (edge.node_1 == node_2 and edge.node_2 == node_1):
-                edge_to_split = edge
+                return edge
+        raise Exception(f"There is no edge between [{node_1}] and [{node_2}]")
 
-        if edge_to_split is None:
-            return None
+    def split_edge(self, node_1: Node, node_2: Node, is_hanging: bool = True) -> Node | None:
```
Comment (strjakub) [inline graph/hypergraph.py:36]: I think it cannot return `None`

**Review by strjakub — APPROVED** [review]

File: graph/hypergraph.py (hunk 6 above)
Comment (Loloxon) [inline graph/hypergraph.py:36]:
```suggestion
    def split_edge(self, node_1: Node, node_2: Node, is_hanging: bool = True) -> Node:
```

**Pull request merged**